glib/src/value.rs

```diff
@@ -56,31 +51,262 @@ use crate::gstring::GString;
 use crate::translate::*;
 use crate::types::{StaticType, Type};
 
-/// An error returned from the [`get`](struct.Value.html#method.get)
-/// or [`get_some`](struct.Value.html#method.get_some) functions on a [`Value`](struct.Value.html)
-#[derive(Clone, Copy, Debug, Eq, PartialEq)]
-pub struct GetError {
-    pub actual: Type,
-    pub requested: Type,
+/// A type that can be stored in `Value`s.
+pub trait ValueType: ToValue + for<'r> FromValue<'r> {
+    /// Type to get the `Type` from.
+    ///
+    /// This exists only for handling optional types.
+    // FIXME: Should default to Self once associated type defaults are stabilized
+    // https://github.com/rust-lang/rust/issues/29661
+    type Type: StaticType;
+}
+
+/// A type that can be stored in `Value`s and is optional.
+///
+/// These are types were storing an `Option` is valid. Examples are `String` and all object types.
+pub trait ValueTypeOptional:
+    ValueType
+    + ToValueOptional
+    + for<'r> FromValue<'r, Error = WrongValueTypeOrNoneError<Self>>
+    + StaticType
+{
+}
+
+impl<'r, T: ValueTypeOptional + StaticType> ValueType for Option<T>
+where
+    Self: for<'s> FromValue<'s, Error = WrongValueTypeOrNoneError<T>>,
+{
+    type Type = T::Type;
+}
+
+/// Trait for errors returned from `Value::get`.
+pub trait GetValueError: std::error::Error + Sized {
+    fn check(value: &Value) -> Result<(), Self>;
 }
 
-impl GetError {
-    pub fn new_type_mismatch(actual: Type, requested: Type) -> Self {
-        GetError { actual, requested }
+/// An error returned from the [`get`](struct.Value.html#method.get) function
+/// on a [`Value`](struct.Value.html) for non-optional types an `Option`.
+pub struct WrongValueTypeError<T: StaticType> {
+    actual: Type,
+    requested: std::marker::PhantomData<*const T>,
+}
+
+unsafe impl<T: StaticType> Send for WrongValueTypeError<T> {}
+unsafe impl<T: StaticType> Sync for WrongValueTypeError<T> {}
+
+impl<T: StaticType> PartialEq for WrongValueTypeError<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.actual == other.actual
     }
 }
 
-impl fmt::Display for GetError {
+impl<T: StaticType> Eq for WrongValueTypeError<T> {}
+
+impl<T: StaticType> GetValueError for WrongValueTypeError<T> {
+    fn check(value: &Value) -> Result<(), Self> {
+        unsafe {
+            if gobject_ffi::g_type_check_value_holds(&value.0, T::static_type().to_glib())
+                == ffi::GFALSE
+            {
+                Err(WrongValueTypeError::type_mismatch(from_glib(
+                    value.0.g_type,
+                )))
+            } else {
+                Ok(())
+            }
+        }
+    }
+}
+
+impl<T: StaticType> WrongValueTypeError<T> {
+    pub fn type_mismatch(actual: Type) -> Self {
```

Comment:
Since this is a constructor, we'd probably prefer implementing `From`.

Reply:
No opinion here, I'll change it unless someone has a good reason not to :)

Reply:
I would allow using using things like `Err(type_expr)?` to return `Result<_, ValueTypeMismatchError>`, which I find neat :)

Reply:
Neat but also a bit confusing :)

Reply:
> Neat but also a bit confusing :)

I thought it was a common pattern. Anyway, I like it.

Reply:
Returning a non-error type (`glib::Type`) via `Err(_)` just seems strange to me :)

Reply:
Well you can see it like this: the error type is `ValueTypeMismatchError` which requires the mismatched type to be constructed, so I return said type and let `rustc` take care about the boilerplates.